lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp

```diff
@@ -1171,59 +1239,243 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
+  }
+
+  void buildEltwiseBinaryPushOutput(Operation &op, OpBuilder &builder,
+                                    ArrayRef<BlockArgument> cbOperands) const {
+    auto outCB = cbOperands[2];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBPushBackOp>(op.getLoc(), outCB, one);
+  }
+
+  void buildEltwiseUnaryReserveInput(Operation &op, OpBuilder &builder,
+                                     ArrayRef<BlockArgument> cbOperands) const {
+    auto inCB = cbOperands[0];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBReserveBackOp>(op.getLoc(), inCB, one);
+  }
+
+  void buildEltwiseUnaryPushInput(Operation &op, OpBuilder &builder,
+                                  ArrayRef<BlockArgument> cbOperands) const {
+    auto inCB = cbOperands[0];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBPushBackOp>(op.getLoc(), inCB, one);
+  }
+
+  void buildEltwiseUnaryWaitInput(Operation &op, OpBuilder &builder,
+                                  ArrayRef<BlockArgument> cbOperands) const {
+    auto inCB = cbOperands[0];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBWaitFrontOp>(op.getLoc(), inCB, one);
+  }
+
+  void buildEltwiseUnaryPopInput(Operation &op, OpBuilder &builder,
+                                 ArrayRef<BlockArgument> cbOperands) const {
+    auto inCB = cbOperands[0];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBPopFrontOp>(op.getLoc(), inCB, one);
+  }
+
+  void buildEltwiseUnaryWaitOutput(Operation &op, OpBuilder &builder,
+                                   ArrayRef<BlockArgument> cbOperands) const {
+    auto outCB = cbOperands[1];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBWaitFrontOp>(op.getLoc(), outCB, one);
+  }
+
+  void buildEltwiseUnaryPopOutput(Operation &op, OpBuilder &builder,
+                                  ArrayRef<BlockArgument> cbOperands) const {
+    auto outCB = cbOperands[1];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBPopFrontOp>(op.getLoc(), outCB, one);
+  }
+
+  void
+  buildEltwiseUnaryReserveOutput(Operation &op, OpBuilder &builder,
+                                 ArrayRef<BlockArgument> cbOperands) const {
+    auto outCB = cbOperands[1];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBReserveBackOp>(op.getLoc(), outCB, one);
+  }
+
+  void buildEltwiseUnaryPushOutput(Operation &op, OpBuilder &builder,
+                                   ArrayRef<BlockArgument> cbOperands) const {
+    auto outCB = cbOperands[1];
+    auto one = i32(1, builder);
+
+    builder.create<ttkernel::CBPushBackOp>(op.getLoc(), outCB, one);
+  }
+
+  void createNoc0DataMovementBlock(
+      ttir::GenericOp genericOp, Block *origGenericOpBlock,
+      SmallVector<Block *> &noc0Blocks, std::int64_t numDPSInputs,
+      ArrayRef<bool> reducedMemrefDims,
+      ArrayRef<StreamedOperand> streamedOperands) const {
+    if (!streamedOperands.empty()) {
```

Comment:
while we are refactoring this, can we do something more about streamedOperands? I don't like that we have this top level distinction 
```
if (!streamedOperands.empty())
...
```

Can we encapsulate every operand in this structure, and we can rename it to OperandDataMovement or similar? Then inside we will have list of movements if needed, otherwise it's empty. 

Then you can call createDataMovementThread for every operand like I did for streamed ones. That function should build reserve op, noc transaction ops if needed, and pushback op. 

I want to eliminate as much as possible notion of op types and embed most of info into this struct that can later help us build any custom data movement kernel. Does this make sense @pjanevskiTT?